src/de/mod.rs

```diff
@@ -421,16 +416,53 @@ where
     write_i32(&mut bytes, length).map_err(Error::custom)?;
 
     reader.take(length as u64 - 4).read_to_end(&mut bytes)?;
+    Ok(bytes)
+}
 
-    let mut deserializer = raw::Deserializer::new(bytes.as_slice());
-    T::deserialize(&mut deserializer)
+/// Deserialize an instance of type `T` from an I/O stream of BSON.
+pub fn from_reader<R, T>(reader: R) -> Result<T>
+where
+    T: DeserializeOwned,
+    R: Read,
+{
+    let bytes = reader_to_vec(reader)?;
+    from_slice(bytes.as_slice())
+}
+
+/// Deserialize an instance of type `T` from an I/O stream of BSON, replacing any invalid UTF-8
+/// sequences with the Unicode replacement character.
+///
+/// This is mainly useful when reading raw BSON returned from a MongoDB server, which
+/// in rare cases can contain invalidly truncated strings (https://jira.mongodb.org/browse/SERVER-24007).
+/// For most use cases, `bson::from_slice` can be used instead.
+pub fn from_reader_utf8_lossy<R, T>(reader: R) -> Result<T>
```

Comment:
per our discussion in slack, I added new functions for the lossy UTF-8 approach.

cc @abr-egn @isabelatkinson